ipsec: Modularization of Agent

smagnani96 opened the pull request:
This PR is the modularization of the IPsec-related bits (agent and bpf map).
Similarly as for the WG modularization (#40360), I kept commit history as clean as possible to ease the review process.

Fixes: #23901

```release-note
Modularization of IPsec Agent.
```

bimmlerd reviewed (approved):
LGTM!

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -1192,6 +1192,7 @@ func (a *Agent) setIPSecSPI(spi uint8) error {
 		a.log.Warn("cilium_encrypt_state map updated failed", logfields.Error, err)
```
> Prior to this commit and modularization, the SPI could've changed at runtime due to a new key being written in the watched file. However, that update would have affected only the underlying BPF map, and not the key custodian (the current IPsecAgent). It is true that the SPI() method of the agent seems to be called only during the daemon startup, but if in future we'd need to call it again and the key changed in the meantime the method
would've returned an outdated SPI value.

So IIUC this is not a bug that can affect users at the moment, right?

smagnani96 replied:
Correct. The method is used only in `newDaemon` to [annotate](https://github.com/cilium/cilium/blob/main/daemon/cmd/daemon.go#L636-L644) the node.
The thing is that subsequent changes to the SPI from the [ipsec job](https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec/ipsec_linux.go#L1221-L1231) would cause misalignment from the SPI in the BPF map (will contain the updated one) and the one stored in the ipsec agent (will contain the old one inferred during startup). No errors as of today though.
Honestly speaking, this is just so that we don't panic if we'll start using `.SPI()` in other places in the future and notice a different value. Maybe this is worth a separate PR + backports?

pchaigno replied:
I think it's fine as is, but up to you. In any case, the separate commit is nice in case we need to backport just that in the future.

smagnani96 replied:
Opened https://github.com/cilium/cilium/pull/41456 for clarity.
Will remove commit in this PR after 🔝  is being merged and this one gets rebased.

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -240,7 +240,7 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
 
```
> With this commit, we also run `make -C Documentation update-cmdref`.

Do you mean if was required on previous commits? Why not just run a `git rebase --exec "make -C Documentation update-cmdref" origin/main` in case someone needs to bisect this stuff someday?

smagnani96 replied:
In theory, that commit is the only one modifying command references, so previous one don't need to run this.
Given I've to solve conflicts anyway, would you prefer me to run this just in case I missed some bits?

pchaigno replied:
I mean, it doesn't hurt and takes 5-10min, right?

smagnani96 replied:
Will do after addressing above comment (rebase).

pchaigno reviewed (approved):
The split into smaller commits will be really valuable for backports :pray: 

Only two questions from me below.

smagnani96 committed:
refactor:ipsec: move fake IPsecAgent under pkg/datapath/fake/types

This commit moves the fake IPsecAgent in pkg/datapath/fake/ipsec.go file
under pkg/datapath/fake/types. While doing that, we now declare the fake
IPsecAgent type as publicly accessible, so that it can be used when needed
a fake datapath type in tests.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
refactor:ipsec: rename IPsecKeyCustodian to IPSecAgent

This commit simply renames the previous IPsecKeyCustodian to IPSecAgent.
This way, we align to the naming being used for the WireGuard agent.
In addition, in subsequent commits we will cellify the IPSecAgent even
further, and so the naming will be more representative.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
refactor:ipsec: move IPSecAgent definition to specific file

This commit moves the IPSecAgent definition to the ipsec_linux.go file.
This is to have everything related to the agent in one place, while leaving
the cell definition and the constructor returning the interface in the cell.go file.
This aligns to the WireGuard agent files structure. However, we do not
rename the file to preserve history.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
refactor:ipsec: move IPSecParameters under pkg/datapath/types

This commit moves IPSecParameters to pkg/datapath/types. This is needed, as in
subsequent commits few methods from the IPSecAgent that accepts such structure
as parameter will be moved in the interface. Therefore, let's declare such
data in the same place.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
modularization:ipsec: move module variables inside IPsecAgent

This commit ensures that all module-wide variable previously used and
shared among different subsystems/callers (also for tests) are now part
of the IPsecAgent.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
refactor:ipsec: make ipsec methods private to the agent

This is a simple cleanup commit to make private all those method that we
don't need to export. In next commits, we will remove the log as parameter
from these methods given they've already access to the agent logger.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
refactor:ipsec: simplify ipsec agent methods by reusing its logger

Most of the methods in the IPsecAgent accepts an external logger
This is usually passed from node.go and ipsec.go, and such logger
is usually enriched with the 'ipsec' subsystem. However, now that the
IPsecAgent has its own logger let's reuse it.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
modularization:ipsec: DaemonConfig as IPsecAgent cell parameter

With this commit, we store a reference to the DaemonConfig local in the
IPsecAgent. In subsequent commits we'll move to a custom IPsecConfig, so
this commit is just to make history a bit clearer and separate contribs.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
refactor:ipsec: move IPsec checks from initEnv to newDaemon

This commit moves IPsec-related checks into the newDaemon function.
This ease subsequent commits, where we'll move ipsec-related parameters
to Cell-specific configuration. In initEnv, we do not have yet the whole
daemon params, so we could not perform such checks. Let's simply move those
at the beginning of newDaemon.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
modularization:ipsec: refactor ipsec EncryptMap to Cell

This commit refactors the encrypt map to a cell. This map was previously
used both from the IPsecAgent and from the Daemon. This is actual not needed,
therefore with this commit such map is being used only from the ipsec
agent. A reference to it is accepted as cell parameter and stored locally.

The EncryptMap now is an interface, and encryptMap implements it. The IPsecAgent
relies on this interface. The only exported methods are Update (already
been used in our agent code), Lookup and UnpinIfExists, which are instead
useful for testing and debugging (see latest commit of this patch series).
Fields inside EncryptKey and EncryptValue are now public so that they can
be used outside and be unmarshalled when calling Lookup.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
modularization:ipsec: add the IPsecAgent.Enabled() method

This commit adds the `Enabled() bool` method to the IPsec agent interface
and implementation. The methods obviously returns true when such feature
is enabled. We add this bits so that in all those places who have already
access to the ipsec agent as a cell dependency there's no need to check
into the daemon configuration anymore, but rather asks the agent directly.
In subsequent commits, we'll add support also to the IPsec-specific config,
and finally move all ipsec-related fields from the daemon config to the new one.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
modularization:ipsec: add IPsecConfig with interface methods

We add a new Config specific for the IPsec cell, which will be later
enriched with all the CLI flags, which will be moved away from DaemonConfig.
This Config, has 4 methods:

1. `Enabled() bool`: similarly as in the previous commit, we add such
    method to the interface so that every cells that needs to check whether
    ipsec is enabled or not can invoke it.
2. `EncryptedOverlayEnabled() bool`: used to check whether encrypted overlay
    is enabled. This way we do not export the variable for safety.
3. `UseCiliumInternalIP() bool`: used to check whether the agent should
    use CiliumInternalIP with IPsec. This way we do not export the variable for safety.
4. `DNSProxyInsecureSkipTransparentModeCheckEnabled() bool`: used to check
    whether dns proxy should skip transparent mode even with IPsec enabled.
    This way we do not export the variable for safety.

Give in the next commit we move all flags to this Config, this commit
serves to keep a clean history and prepare the infra for that.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
tests:ipsec: add cell tests for IPsecAgent

As a final commit of this patch series, here we introduce tests for the
whole IPsecAgent cell. The tests cover the initialization of the cell,
both when ipsec is enabled and disabled. The tests also verify that the
background tasks are correctly created and working, while also the SPI
is being loaded from the specified file.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
modularization:ipsec move params from DaemonConfig to IPsecConfig

In previous commits, we moved away from using ipsec-related parameters
stored in the DaemonConfig struct, and instead we introduced a new
IPsecConfig struct that contains those parameters and exports them as
methods via interface. In this commit, we complete the modularization,
by finally movind these parameters away from DaemonConfig into the
IPsecConfig, being declared as cell parameters.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>
